src/deploy/functions/prepare.ts

```diff
@@ -495,3 +498,41 @@ export async function loadCodebases(
   }
   return wantBuilds;
 }
+
+// Genkit almost always requires an API key, so warn if the customer is about to deploy
+// a function and doesn't have one. To avoid repetitive nagging, only warn on the first
+// deploy of the function.
+export async function warnIfNewGenkitFunctionIsMissingSecrets(
+  have: backend.Backend,
+  want: backend.Backend,
+  options: DeployOptions,
+) {
+  if (options.force) {
+    return;
+  }
+
+  const newAndMissingSecrets = backend.allEndpoints(
+    backend.matchingBackend(want, (e) => {
+      if (!backend.isCallableTriggered(e) || !e.callableTrigger.genkitAction) {
+        return false;
+      }
+      if (e.secretEnvironmentVariables?.length) {
+        return false;
+      }
+      return !backend.hasEndpoint(have)(e);
+    }),
+  );
+
+  if (newAndMissingSecrets.length) {
+    const msgBase =
+      `The function(s) ${newAndMissingSecrets.map((e) => e.id).join(", ")} use Genkit but do not have access to a secret. ` +
+      "This may cause the function to fail if it depends on an API key. To learn more about granting a function access to " +
+      "secrets, see https://firebase.google.com/docs/functions/config-env?gen=2nd#secret_parameters.";
+    if (options.nonInteractive) {
+      throw new FirebaseError(msgBase);
+    }
+    if (!(await prompt.confirm({ message: `${msgBase} Continue?` }))) {
+      throw new FirebaseError("Aborted");
+    }
+  }
```

Comment:
nit* prompts.confirm has logic to short-circuit if nonInteractive is passed. So could be re-written to:

```ts
    logLabledBullet(msgBase);
    if (!(await prompt.confirm({ nonInteractive: options.nonInteractive, force: options.force }))) {
      throw new FirebaseError("Aborted");
    }
```

Slightly different behavior in messaging.